### ISSUE TITLE
Poor mIoU when training on 1464 images in a supervised manner

### ISSUE THREAD

**Issue by xiaomengyc** [yassouali/CCT#10]
Hi,
I trained the model on the 1464 images in a supervised manner. The highest mIoU on Val is 67.00%, but the reported number in your paper is 69.4%.  Here is my `config.json` file. Can you have a look at which part is wrong?
```javascript
 {  "name": "CCT",
    "experim_name": "CCT",
    "n_gpu": 1,
    "n_labeled_examples": 1464,
    "diff_lrs": true,
    "ramp_up": 0.1,
    "unsupervised_w": 30,
    "ignore_index": 255,
    "lr_scheduler": "Poly",
    "use_weak_lables":false,
    "weakly_loss_w": 0.4,
    "pretrained": true,
    "model":{
        "supervised": true,
        "semi": false,
        "supervised_w": 1,

        "sup_loss": "CE",
        "un_loss": "MSE",

        "softmax_temp": 1,
        "aux_constraint": false,
        "aux_constraint_w": 1,
        "confidence_masking": false,
        "confidence_th": 0.5,

        "drop": 6,
        "drop_rate": 0.5,
        "spatial": true,
    
        "cutout": 6,
        "erase": 0.4,
    
        "vat": 2,
        "xi": 1e-6,
        "eps": 2.0,

        "context_masking": 2,
        "object_masking": 2,
        "feature_drop": 6,

        "feature_noise": 6,
        "uniform_range": 0.3
    },


    "optimizer": {
        "type": "SGD",
        "args":{
            "lr": 1e-2,
            "weight_decay": 1e-4,
            "momentum": 0.9
        }
    },


    "train_supervised": {
        "data_dir": "../data/VOC2012",
        "batch_size": 10,
        "crop_size": 320,
        "shuffle": true,
        "base_size": 400,
        "scale": true,
        "augment": true,
        "flip": true,
        "rotate": false,
        "blur": false,
        "split": "train_supervised",
        "num_workers": 8
    },

    "train_unsupervised": {
        "data_dir": "VOCtrainval_11-May-2012",
        "weak_labels_output": "pseudo_labels/result/pseudo_labels",
        "batch_size": 10,
        "crop_size": 320,
        "shuffle": true,
        "base_size": 400,
        "scale": true,
        "augment": true,
        "flip": true,
        "rotate": false,
        "blur": false,
        "split": "train_unsupervised",
        "num_workers": 8
    },

    "val_loader": {
        "data_dir": "../data/VOC2012",
        "batch_size": 1,
        "val": true,
        "split": "val",
        "shuffle": false,
        "num_workers": 4
    },

    "trainer": {
        "epochs": 80,
        "save_dir": "saved/",
        "save_period": 5,
  
        "monitor": "max Mean_IoU",
        "early_stop": 10,
        
        "tensorboardX": true,
        "log_dir": "saved/",
        "log_per_iter": 20,

        "val": true,
        "val_per_epochs": 5
    }
}
```

**Comment by yassouali** [comment]
Hi xiaomengyc;

Thank you for you interest in our work.

As for the results:
![image](https://user-images.githubusercontent.com/45082182/91548696-0deebc00-e926-11ea-8c00-70b5f7d1ac0b.png)

69.4% is obtained with our method CCT, so in a semi-supervised mode (`semi:true`), in your case, you are only training with a supervised loss on the 1.5K examples without CCT loss, so its expected to have worse performance. The config provided gives the results provided in the paper.

**Comment by xiaomengyc** [comment]
Hi yassouali,
Thanks for your reply!

I found the problem. It needs to be tested in a multiscale manner as the inference script does. I got 69.06 when using multi-scale testing.

**Comment by yassouali** [comment]
Hi @xiaomengyc 

Just to emphasize, the results above are not done in multiscale inference, if you want to reproduce them, use the mode `semi:true`, the supervised mode is only used to make sure the performances are better with semi mode.

If you do multiscale inference with semi mode, the expected results are in the appendix, as follows:

![image](https://user-images.githubusercontent.com/45082182/91549408-30350980-e927-11ea-92ca-fd0c1f8c9c1d.png)

**Comment by xiaomengyc** [comment]
I misunderstood your method. I thought the CCT augmentation and loss function are only conducted on the images without pixel-level masks. Now, I see your point. The reported 69.4 is got by using abCE loss and CCT loss on the 1.5k images without other images involved.

Thank you so much!

**Comment by yassouali** [comment]
Yes (kinda), so 69.4 with n=1.5K means that we trained with a normal CE on the 1.5 examples, and CCT loss on the rest (9K).

Hope I answered your questions.

**Comment by yassouali** [comment]
I'll close this issue for now, if you have any other questions, we can reopen it.

**Comment by won-bae** [comment]
> Yes (kinda), so 69.4 with n=1.5K means that we trained with a normal CE on the 1.5 examples, and CCT loss on the rest (9K).
> 
> Hope I answered your questions.

Hi, thanks for the explanation. But, I am still confused about the config.
With the exactly same config and CCT_1500.pth checkpoint you uploaded, I got 69.4%. If my understanding is correct, that corresponds to normal CE on 1.5K and CCT loss on 9K. Then how do I just use normal CE on 1.5K without using CCT loss and get 67.xx%? Thank you in advance!

**Comment by yassouali** [comment]
Hi @won-bae 

Thank you for your interest;

This ca be done using the config file, simply set `supervised: true,` and `semi: false` and your good to go.

**Comment by won-bae** [comment]
I see. Thank you so much for your help @yassouali
It took me sometime to actually train a model. 
One last help is needed. For `supervised: false`, `semi: true` and `use_weak_lables: false`, it corresponds to CCT with 1.5K pixel labels (without any weak labels), right?

**Comment by yassouali** [comment]
Yes, in this case, you'll train using only 1.5K labels with the CE loss only

**Comment by won-bae** [comment]
Thank you for the clarification but weirdly, I get the performance as high as 70.2% although you reported 69.4%. I think the gap is too large even considering the randomness. Here is the config I used. Could you point me out which config is wrong if there is any?

```
{
    "diff_lrs": true,
    "experim_name": "CCT",
    "ignore_index": 255,
    "lr_scheduler": "Poly",
    "model": {
        "aux_constraint": false,
        "aux_constraint_w": 1,
        "confidence_masking": false,
        "confidence_th": 0.5,
        "context_masking": 2,
        "cutout": 6,
        "drop": 6,
        "drop_rate": 0.5,
        "eps": 2.0,
        "erase": 0.4,
        "feature_drop": 6,
        "feature_noise": 6,
        "object_masking": 2,
        "semi": true,
        "softmax_temp": 1,
        "spatial": true,
        "sup_loss": "CE",
        "supervised": false,
        "supervised_w": 1,
        "un_loss": "MSE",
        "uniform_range": 0.3,
        "vat": 2,
        "xi": 1e-06
    },
    "n_gpu": 1,
    "n_labeled_examples": 1464,
    "name": "CCT",
    "optimizer": {
        "args": {
            "lr": 0.01,
            "momentum": 0.9,
            "weight_decay": 0.0001
        },
        "type": "SGD"
    },
    "pretrained": true,
    "ramp_up": 0.1,
    "train_supervised": {
        "augment": true,
        "base_size": 400,
        "blur": false,
        "crop_size": 320,
        "data_dir": "VOCtrainval_11-May-2012",
        "flip": true,
        "ignore_index": 255,
        "mean": [
            0.485,
            0.456,
            0.406
        ],
        "n_labeled_examples": 1464,
        "rotate": false,
        "scale": true,
        "split": "train_supervised",
        "std": [
            0.229,
            0.224,
            0.225
        ]
    },
    "train_unsupervised": {
        "augment": true,
        "base_size": 400,
        "blur": false,
        "crop_size": 320,
        "data_dir": "VOCtrainval_11-May-2012",
        "flip": true,
        "ignore_index": 255,
        "mean": [
            0.485,
            0.456,
            0.406
        ],
        "n_labeled_examples": 1464,
        "rotate": false,
        "scale": true,
        "split": "train_unsupervised",
        "std": [
            0.229,
            0.224,
            0.225
        ],
        "use_weak_lables": false,
        "weak_labels_output": "pseudo_labels/result/pseudo_labels"
    },
    "trainer": {
        "early_stop": 10,
        "epochs": 80,
        "log_dir": "saved/",
        "log_per_iter": 20,
        "monitor": "max Mean_IoU",
        "save_dir": "/home/whbae/scratch/CCT.tmp/1464_semi",
        "save_period": 5,
        "tensorboardX": true,
        "val": true,
        "val_per_epochs": 5
    },
    "unsupervised_w": 30,
    "use_weak_lables": false,
    "val_loader": {
        "data_dir": "VOCtrainval_11-May-2012",
        "ignore_index": 255,
        "mean": [
            0.485,
            0.456,
            0.406
        ],
        "split": "val",
        "std": [
            0.229,
            0.224,
            0.225
        ],
        "val": true
    },
    "weakly_loss_w": 0.4
}
```
